Tests/CacheAdvanceTests/CacheAdvanceTests.swift

```diff
@@ -97,34 +97,60 @@ final class CacheAdvanceTests: XCTestCase {
     }
 
     func test_messages_throwsFileCorruptedWhenOffsetInFileAtEndOfNewsetMessageOutOfSync() throws {
```

Comment:
Given the above, I would like to better understand how the file corruption you believe you are experiencing could occur. That said, I do still think it is valuable to introduce checks to avoid the infinite loop problem on big-endian machines.

Reply:
Agree about the this. 

> Since [iOS runs in little-endian mode](https://bignerdranch.com/blog/its-the-endian-of-the-world-as-we-know-it/), and the issue you're encountering is occurring on iOS, I do not think we've as-of-yet identified how file corruption would occur.

This is my first time go deep with the integer representation in a real code. 

In little-endian mode, it should be possible to cause the number to be smaller than the last value when crashed when writing.

The`offsetInFileAtEndOfNewestMessage` to write is 
`00002000`. 
The old value in the header is 
`00101000`.
If the write action crash when only the first bytes in `00002000` are written, 
Then we got 
`00001000` 
which is smaller than `00101000` in little-endian mode. 

From this, I added another test with smaller value in the header, it runs into the infinite loop too.
`  func test_messages_timeout_WhenOffsetInFileAtEndOfNewestMessageSmallThanActualFileLength() throws {
`
@dfed

Reply:
`00002000` is not valid binary: the bits written to disk are only `1`s and `0`s`.

After reviewing the test, I see that we enter an infinite loop because the `offsetInFileAtEndOfNewestMessage` is corrupted! Good catch. I was wrong when I wrote:

> I think in this case we would likely end up reading a [partial message for the last message](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheReader.swift#L48), which would result in reading data that would [fail to decode](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L174), leading to the `messages()` call throwing a decoding error

That statement was wrong because we don't validate that the `messageLength` in `.span(messageLength)` does not exceed `offsetInFileAtEndOfNewestMessage`!! We should fix this @jianjunwoo

Reply:
> We should fix this
Agree, we could fix this so that the infinite loop should never happen.
But, I still want a parameter for experiment :D . Let me have a fork repo do this. Do you think it's fine? @bachand

Reply:
Given all of the above, I think we have identified that we should make the following changes:
1. Add a check above [this line](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheReader.swift#L55) that checks to ensure our read `messageLength` is valid given what we know about the `offsetInFileAtEndOfNewestMessage` and `maximumBytes`. This `guard` would look something like `(startingOffset < offsetInFileAtEndOfNewestMessage && startingOffset + messageLength <= offsetInFileAtEndOfNewestMessage) || startingOffset >= offsetInFileAtEndOfNewestMessage && startingOffset + messageLength < maximumBytes` (though we should likely write this over multiple lines to make it more clear). If this guard fails, we'll `throw CacheAdvanceError.fileCorrupted`.
2. Add a new `guard` [above this line](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheReader.swift#L63) that the current cache overwrites new messages. If that `guard` does not pass, we should `throw CacheAdvanceError.fileCorrupted`
3. Add a check to the `CacheReader`'s `nextEncodedMessage(...)` method that ensures we don't seek to the beginning of the file more than once.

These changes can be done in the same PR or separate PRs.

As an aside, I think there's an open question regarding whether we want to attach the messages we were able to read before detecting file corruption to the thrown error. Today we don't, but that could be a nice improvement. This kind of improvement could be made on a separate PR as well.

Reply:
If you'd like, I'd be happy to write the PRs that implement the above checks. I'm also happy to review PRs that accomplish the above. You + Zack already done a great job of finding issues in this library: your reward for that can be reviewing a PR I write rather than having to do the work yourself 😉. Let me know what you have time to contribute and where I can help!

Reply:
Looking forward to the checks working! Go ahead pls : ) 

I would like to fork one, and have a quick fix for the treatment group.
The `shouldOverwriteOldMessages` is false in Airbnb, so it should be safe to apply my quick fix.

Reply:
Working on implementing the above. I think I've found some simpler checks than what I outlined above. PR incoming once I iron out a few more tests 😄

Reply:
I believe we can close this PR in favor of #66, as that PR addresses the issues discovered here.

Reply:
Regarding

> If we crash during step 1, the the next time we open the file we could start reading at an effectively random point in the file. It is highly likely that any "message" read in this file would [fail to decode](https://github.com/dfed/CacheAdvance/blob/0e63e448850a0d64a2a228d59d293749c3c1a248/Sources/CacheAdvance/CacheAdvance.swift#L174), leading to the `messages()` call throwing a decoding error. A crash during step 1 would therefore not result in an infinite reading loop.

it's possible for someone to write a custom message encoder/decoder. If someone were to write a message encoder/decoder that wrote/read raw data, then the encoding would never fail.

Does this change our opinion about the risk of an infinite loop with (1)?

Reply:
If the decoder never fails, then we have to consider the following failure cases:
1. The crash occurred while writing the message's length (e.g. the `messageSpan`)
2. The crash occurred while writing the message's content (e.g. the `data`)

If the message length didn't write correctly, then we'll end up reading effectively random values for the message length on future "messages". We'll eventually hit a case I've outlined in #66 where a message's length ends after `offsetInFileAtEndOfNewestMessage` and we throw a corrupted file error.

If the message data didn't write correctly, then we'll send the bad message to consuming application, but I don't see how bad message data would cause us to infinite loop.